cmd/notary/integration_test.go

```diff
@@ -1526,3 +1527,447 @@ func TestWitness(t *testing.T) {
+	output, err = runCommand(t, tempDir, "-s", server.URL, "delegation", "list", "gun")
+	require.NoError(t, err)
+	require.Contains(t, output, "path")
+	require.Contains(t, output, keyID)
+	require.Contains(t, output, keyID2)
+
+	// remove the delegation's first key
+	output, err = runCommand(t, tempDir, "delegation", "-s", server.URL, "remove", "-p", "gun", "targets/delegation", keyID)
+	require.NoError(t, err)
+	require.Contains(t, output, "Removal of delegation role")
+
+	// list delegations - we should see the delegation but with only the second key
+	output, err = runCommand(t, tempDir, "-s", server.URL, "delegation", "list", "gun")
+	require.NoError(t, err)
+	require.NotContains(t, output, keyID)
+	require.Contains(t, output, keyID2)
+
+	// remove the delegation's second key
+	output, err = runCommand(t, tempDir, "delegation", "-s", server.URL, "remove", "-p", "gun", "targets/delegation", keyID2)
+	require.NoError(t, err)
+	require.Contains(t, output, "Removal of delegation role")
+
+	// list delegations - we should see no delegations
+	output, err = runCommand(t, tempDir, "-s", server.URL, "delegation", "list", "gun")
+	require.NoError(t, err)
+	require.NotContains(t, output, keyID)
+	require.NotContains(t, output, keyID2)
+}
+
+// TestClientTUFAddByHashWithAutoPublish is similar to TestClientTUFAddByHashInteraction,
+// but with the auto publish flag "-p".
+func TestClientTUFAddByHashWithAutoPublish(t *testing.T) {
+	// -- setup --
+	setUp(t)
+
+	tempDir := tempDirWithConfig(t, "{}")
+	defer os.RemoveAll(tempDir)
+
+	server := setupServer()
+	defer server.Close()
+
+	targetData := []byte{'a', 'b', 'c'}
+	target256Bytes := sha256.Sum256(targetData)
+	targetSha256Hex := hex.EncodeToString(target256Bytes[:])
+	target512Bytes := sha512.Sum512(targetData)
+	targetSha512Hex := hex.EncodeToString(target512Bytes[:])
+
+	err := ioutil.WriteFile(filepath.Join(tempDir, "tempfile"), targetData, 0644)
+	require.NoError(t, err)
+
+	var (
+		output  string
+		target1 = "sdgkadga"
+		target2 = "asdfasdf"
+		target3 = "qwerty"
+	)
+	// -- tests --
+
+	// init repo
+	_, err = runCommand(t, tempDir, "-s", server.URL, "init", "gun", "-p")
+	require.NoError(t, err)
+
+	// add a target just by sha256
+	_, err = runCommand(t, tempDir, "-s", server.URL, "addhash", "-p", "gun", target1, "3", "--sha256", targetSha256Hex)
+	require.NoError(t, err)
+
+	// check status - no targets
+	output, err = runCommand(t, tempDir, "status", "gun")
+	require.NoError(t, err)
+	require.False(t, strings.Contains(string(output), target1))
+
+	// list repo - see target
+	output, err = runCommand(t, tempDir, "-s", server.URL, "list", "gun")
+	require.NoError(t, err)
+	require.True(t, strings.Contains(string(output), target1))
+
+	// lookup target and repo - see target
+	output, err = runCommand(t, tempDir, "-s", server.URL, "lookup", "gun", target1)
+	require.NoError(t, err)
+	require.True(t, strings.Contains(string(output), target1))
+
+	// remove target
```

Comment:
Totally non-blocking - since this test case is just testing the autopublish functionality, the removal and testing different types of hashes may not be as necessary.